### DISCUSSION TITLE
Add an option to e-mail fields to allow internationalized domain names

### DISCUSSION THREAD

**Discussion by MoritzLost** [craftcms/cms#9698] [Ideas] [answered]
Craft's built-in e-mail field, at the moment, does not support international e-mails and there's no straightforward way to modify that behaviour – at least as far as I can tell. An option to allow international emails in the field settings would be a great addition to the field.

Below is a bit of background and my current workaround for this issue.

---

Craft's email field uses [Yii2 EmailValidator class](https://www.yiiframework.com/doc/api/2.0/yii-validators-emailvalidator). This class has an `$enableIDN` property to allow internationalized domain names. Craft's email field [doesn't use that option, though](https://github.com/craftcms/cms/blob/main/src/fields/Email.php#L98-L107).

I haven't found a straightforward way to modify the rules returned by a field's `getElementValidationRules` method – is there one? The only thing I can think of is extending the built-in E-Mail field to modify the validation rules, but that seems a bit overkill.

Anyway, I managed to allow international e-mails using the hook [craft\base\Element::EVENT_DEFINE_RULES](https://docs.craftcms.com/api/v3/craft-base-model.html#event-define-rules) in a site module (full code below). But that's a bit of a brute-force approach and I'm not sure if it handles all edge cases. So having either a method to override element validation rules for a single field in a straightforward way would be great.

How about adding an option to the e-mail field to allow internationalized domain names? If this were to be added, there should probably be an additional option to allow/prevent IDN characters in the local part of the email as well, as per [this issue](https://github.com/yiisoft/yii2/issues/18734).

```php
<?php

namespace modules\Site;

use Craft;
use craft\base\Element;
use craft\events\DefineRulesEvent;
use yii\base\Event;

class SiteModule extends \yii\base\Module
{
    public function init()
    {
        Event::on(
            Element::class,
            Element::EVENT_DEFINE_RULES,
            [$this, 'allowIDNEmailsForElements']
        );

        parent::init();
    }

    public function allowIDNEmailsForElements(DefineRulesEvent $e)
    {
        /** @var craft\base\Element $element */
        $element = $e->sender;

        foreach ($e->rules as $index => $rule) {
            if (!empty($rule[1]) && $rule[1] === 'email') {
                $e->rules[$index]['enableIDN'] = true;
            }
        }
    }
}
```

**Comment by brandonkelly** [comment]
We recently added support for this to user account emails, and probably should have done it for Email fields at the same time, but they were overlooked. So will look into getting this into the next release.

**Reply by MoritzLost** [reply]
Awesome, thank you!

**Comment by brandonkelly** [comment, marked as answer]
Craft 3.7.9 is out now with support for IDNA ASCII characters in Email field values ✨

**Reply by MoritzLost** [reply]
Awesome, thanks!

**Reply by thinktrans** [reply]
Awesome to see email implementations being more and more aware about the Internationalized Domain Names and the Email IDs. If I may, I would request the Developers to see the analysis presented by the Universal Acceptance Steering Group about the conventions/best practices to be followed while naming/allowing the local parts of the Email field. Here is the [link](https://uasg.tech/download/uasg-028-considerations-for-naming-internationalized-email-mailboxes-en/).

Thank you for being one of the early adopters of IDNs and EAI.  :)
